watertap/property_models/activated_sludge/modified_asm2d_reactions.py

```diff
@@ -10,81 +10,1592 @@
+            ("R19", "Liq", "S_N2"): 0,
+            ("R19", "Liq", "S_NO3"): 0,
+            ("R19", "Liq", "S_PO4"): -(
+                self.f_XI * self.i_PXI + (1 - self.f_XI) * self.i_PXS - self.i_PBM
+            ),
+            ("R19", "Liq", "S_IC"): -(
+                self.f_XI * self.i_CXI + (1 - self.f_XI) * self.i_CXS - self.i_CXB
+            ),
+            ("R19", "Liq", "X_I"): self.f_XI,
+            ("R19", "Liq", "X_S"): 1 - self.f_XI,
+            ("R19", "Liq", "X_H"): 0,
+            ("R19", "Liq", "X_PAO"): 0,
+            ("R19", "Liq", "X_PP"): 0,
+            ("R19", "Liq", "X_PHA"): 0,
+            ("R19", "Liq", "X_AUT"): -1,
+            ("R19", "Liq", "S_K"): 0,
+            ("R19", "Liq", "S_Mg"): 0,
+        }
+        # Fix all the variables we just created
+        for v in self.component_objects(pyo.Var, descend_into=False):
+            v.fix()
+
+    @classmethod
+    def define_metadata(cls, obj):
+        obj.add_properties(
+            {
+                "reaction_rate": {"method": "_rxn_rate"},
+            }
+        )
+        obj.add_default_units(
+            {
+                "time": pyo.units.s,
+                "length": pyo.units.m,
+                "mass": pyo.units.kg,
+                "amount": pyo.units.kmol,
+                "temperature": pyo.units.K,
+            }
+        )
+
+
+class _ModifiedASM2dReactionBlock(ReactionBlockBase):
+    """
+    This Class contains methods which should be applied to Reaction Blocks as a
+    whole, rather than individual elements of indexed Reaction Blocks.
+    """
+
+    def initialize(self, outlvl=idaeslog.NOTSET, **kwargs):
+        """
+        Initialization routine for reaction package.
+
+        Keyword Arguments:
+            outlvl : sets output level of initialization routine
+
+        Returns:
+            None
+        """
+        init_log = idaeslog.getInitLogger(self.name, outlvl, tag="properties")
+        init_log.info("Initialization Complete.")
 
 
 @declare_process_block_class(
     "ModifiedASM2dReactionBlock", block_class=_ModifiedASM2dReactionBlock
 )
-class ModifiedASM2dReactionBlockData(ASM2dReactionBlockData):
-    pass
+class ModifiedASM2dReactionBlockData(ReactionBlockDataBase):
+    """
+    Reaction Block for ASM2d.
+    """
+
+    def build(self):
+        """
+        Callable method for Block construction
+        """
+        super().build()
+
+        # Create references to state vars
+        # Concentration
+        add_object_reference(self, "conc_mass_comp_ref", self.state_ref.conc_mass_comp)
+
+    # Rate of reaction method
+    def _rxn_rate(self):
+        self.reaction_rate = pyo.Var(
+            self.params.rate_reaction_idx,
+            initialize=0,
+            doc="Rate of reaction",
+            units=pyo.units.kg / pyo.units.m**3 / pyo.units.s,
+        )
+
+        try:
+            # TODO: Refer to Flores-Alsina c code to account for sulfur in rate expressions
```

Comment:
Since `S_K` and `S_Mg` are added as reactants, shouldn't they be accounted in these rate expressions? Or they are just zero-order reactants?

Reply:
S_K and S_Mg were not implemented in their version of the rate expression, so I assume they have no effect (zero-order reactant). If you want to take a look at their code, there's a Box link in the PR header. It might be a bit overwhelming to look at at first, but the rate equations are written as `proc## = rate expression` and the terms corresponding to S_K and S_Mg are `xtemp[16]` and `xtemp[17]`, respectively. Neither of which are included in any of the rate expressions.

Reply:
OK. That make sense!